### ISSUE TITLE
HashSet serialization problem with contains(pojo)

### ISSUE THREAD

**Issue by Bogdan-Vasile** [protostuff/protostuff#231]
When deserializing a HashSet<Pojo>, the result will contain the same entities, but the result of calling myHashSet.contains(pojoFromHashSet) will almost never result true.

I.e.: Most of the time the output of the following code will print that the hashSet doesn't contain the respective pojo:

```
HashSet<MyPojo> myDeserializedHashSet = new HashSet<>();
GraphIOUtil.mergeFrom(binaryData, 0, binaryData.length, myDeserializedHashSet, mySchema);
for(MyPojo myPojo : myDeserializedHashSet  ){
    if(myDeserializedHashSet.contains(myPojo)){
        System.out.println("myDeserializedHashSet  contains pojo with hashCode " +myPojo.hashCode() );
    }else{
        System.out.println("myDeserializedHashSet  DOES NOT contain pojo with hashCode " +myPojo.hashCode() );
    }
}
```

A similar post from 2015 reported this error: [https://groups.google.com/forum/#!topic/protostuff/Ub9rQkoqCPc](https://groups.google.com/forum/#!topic/protostuff/Ub9rQkoqCPc)

The problem occurs at least with Pojos that have a custom HashCode implementation, but maybe some other entity types have problems.

Here is what I saw that it happens when deserializing the HashSet:

1.  A collection (HashSet) is detected in the binary input and a new instance is created.

2. The individual elements start being deserialized in the following code (snippet from IdStrategy, around line 335):

```
final DerivativeSchema POLYMORPHIC_POJO_ELEMENT_SCHEMA = new DerivativeSchema(
            this)
    {
        @Override
        @SuppressWarnings("unchecked")
        protected void doMergeFrom(Input input, Schema<Object> derivedSchema,
                Object owner) throws IOException
        {
            final Object value = derivedSchema.newMessage();

            if (MapWrapper.class == owner.getClass())
                ((MapWrapper<Object, Object>) owner).setValue(value);
            else
                ((Collection<Object>) owner).add(value);

            if (input instanceof GraphInput)
            {
                // update the actual reference.
                ((GraphInput) input).updateLast(value, owner);
            }

            derivedSchema.mergeFrom(input, value);
        }
    };

```
The problem with the above code is that immediately after the value creation, the element is added to the collection. This way, the pojo can't have its HashCode correctly calculated because its fields are not yet deserialized. When that pojo is added to the HashSet, its position in the HashSet will be calculated based on its (current) hashCode. After adding it, the pojo's fields will be deserialized as usual, but the future calls to pojo.hashCode() will return a different value, which is the correct one.

After the deserialization is finished, you will have trouble calling myHashSet.contains(aPojoFromMyHashSet).

In order to fix this simple case I saw that it is sufficient to move the following code:
```
            if (MapWrapper.class == owner.getClass())
                ((MapWrapper<Object, Object>) owner).setValue(value);
            else
                ((Collection<Object>) owner).add(value);

```
after
```         
derivedSchema.mergeFrom(input, value);
```
in order to give a chance to the pojo to report the correct hashCode when asked by the HashSet addition operation.

BUT...

The proposed fix works only in case of simple (non-circular) object graphs. In the case of circular object graphs, I saw that the problem persists. I suppose it has to do with the visited references held in the GraphByteArrayInput and different HashSet instances held in a circular graph that contain the same entity. I'll see if I can make a test to reproduce the issue.

I found a fix for this case also, but I'm not happy with it because it will cause a performance drop in the deserialization:
1. Add the following function in the GraphByteArrayInput:

```
        @SuppressWarnings({ "unchecked", "rawtypes" })
	public void fixMessages() {
		ArrayList<Object> tempHolder = new ArrayList<Object>();
		for (Object currentReference : references) {
			if (currentReference instanceof HashSet) {
				HashSet clonedHashSet = (HashSet) currentReference;
				tempHolder.clear();
				tempHolder.addAll(clonedHashSet);
				clonedHashSet.clear();
				clonedHashSet.addAll(tempHolder);
			}
		}
	}

```
2. Modify the following function in the GraphIOUtil (around line 44):
```
    public static <T> void mergeFrom(byte[] data, int offset, int length, T message,
            Schema<T> schema)
    {
        try
        {
            final ByteArrayInput input = new ByteArrayInput(data, offset, length, true);
            final GraphByteArrayInput graphInput = new GraphByteArrayInput(input);
            schema.mergeFrom(graphInput, message);
            input.checkLastTagWas(0);
            graphInput.fixMessages();
        }
        catch (ArrayIndexOutOfBoundsException ae)
        {
            throw new RuntimeException("Truncated.", ProtobufException.truncatedMessage(ae));
        }
        catch (IOException e)
        {
            throw new RuntimeException("Reading from a byte array threw an IOException (should " +
                    "never happen).", e);
        }
    }
```

This calls graphInput.fixMessages() function after finishing the deserialization, and all the deserialized hashSets will be reconstructed.

I said I don't like it because the proposed fix will recreate ALL the HashSets, instead of only the ones with problems, which will cause a performance degradation. Also, if there is no HashSet in the object graph, the code will still iterate trough all the deserialized objects, so maybe it would be best if we keep track of only the deserialized hashSets.

What do you think it would be best way to fix this problem?

**Comment by dyu** [comment]
Thanks for the report and potential fix.
Can you attach isolated samples/test-cases that triggers the issue for both core and graph format?

> I suppose it has to do with the visited references held in the GraphByteArrayInput and different HashSet instances held in a circular graph that contain the same entity. 

If the entity was derialized before it was added to the set, it should address both circular/non-circular datasets.  This is the only thing that need to be fixed for it to work.

What version are you using?
Does your dataset have a pojo that extends HashSet?

**Comment by Bogdan-Vasile** [comment]
Thank you for the fast reply and congratulations for writing such a nice piece of software ( Should have written this in the original message, but I hope you will forgive me :) ).

I'm currently using version 1.5.0, but I think that the issue will show up on all the versions.

The dataset contains only standard HashSets, along with other entities and standard collections.

From my tests I saw that the simple case of serializing a HashSet works by only modifying the doMergeFrom function in POLYMORPHIC_POJO_ELEMENT_SCHEMA as above, and it should be very easy to test.

For the cyclic graph problem, with the entities that I have in my model, I saw that the problem was not fixed with that modification only. Providing a standardized test will be a bit complicated, because I don't know yet the condition for reproducing the graph error and I'm also busy with some projects. 

Nevertheless, I must do it, and I would really appreciate it if you could provide me some hints:
1. In which project should I write a test?
2. What test class would be easier to copy from? I'm currently thinking about RuntimeGraphTest and PolymorphicRuntimeGraphTest.

**Comment by dyu** [comment]
> Thank you for the fast reply and congratulations for writing such a nice piece of software ( Should have written this in the original message, but I hope you will forgive me :) ).

Thanks ... and no worries :-)

> For the cyclic graph problem, with the entities that I have in my model, I saw that the problem was not fixed with that modification only. Providing a standardized test will be a bit complicated, because I don't know yet the condition for reproducing the graph error and I'm also busy with some projects.

Please try at least version 1.5.3 (or the latest 1.5.x) with the 1st fix you mentioned and report back here the results.

> Providing a standardized test will be a bit complicated, because I don't know yet the condition for reproducing the graph error and I'm also busy with some projects.

> Nevertheless, I must do it

Yea you should :-)  That would make it easier/faster for the devs here to make a fix.

> In which project should I write a test?

Easiest would be to pick one test (like you mentioned) and put your test-case in there.  Then you can attach the diff here.

Or you could create a standalone project (on your github repo) with a ```main``` in it

**Comment by Bogdan-Vasile** [comment]
Hi,

I created a test project for HashSet serialization that you can find in the attached zip, along with a patch file for protostuff 1.6.1-SNPSHOT. I thought that this was the easiest way to test against multiple protostuff versions.
[protostuff-HashSetTest.zip](https://github.com/protostuff/protostuff/files/1379185/protostuff-HashSetTest.zip)

During the test creation I saw that the original change that I proposed needed some work, and depending on the serialized classes and environment parameters, more adjustments were necessary.

So I created a few classes for serialization, with and without tag annotations, with and without cycles and with and without a hierarchy. I don't think I managed to create all the possible permutations, but this should be a (good) start. All the classes that are added to the HashSets have custom equals/hashCode implemented, based on the entity Id.

If you want to run the files as tests, you can put them in the protostuff-runtime-registry project, where the other tests are located. You should also uncomment the RuntimeHashSetTestSystemPropertiesDefault and RuntimeHashSetTestSystemPropertiesOverride classes definitions to extend AbstractTest. (The RuntimeHashSetTestSystemPropertiesOverride contain the system properties that I use in my projects, in order to be able to serialize entire graphs of entities).

I created a recursive entities parser (Checker) that inspects recursively a given object for HashSet errors. I didn't inspect the possibility of recursively checking if two objects are equal, partly because the objects seemed to be deserialized correctly.

Most of the errors were solved by modifying the GraphInput to accept a reference to a HashSet and by modifying POLYMORPHIC_POJO_ELEMENT_SCHEMA as in the first message.

I wasn't able to solve the following tests: test_CyclicGraph_DefaultRuntimeSchema_NoHierarchy_Reverse_NoTag
test_CyclicGraph_DefaultRuntimeSchema_NoHierarchy_Tag99

These functions try to serialize a Pojo that has the Id field declared after the Data field. I couldn't apply the same fix as in other cases because I don't have access to the Input when the owner collection is initialized (in RuntimeRepeatedFieldFactory->createCollectionObjectV(...)->setValue(...).

Please ask me for any clarifications (I know I didn't provide much info) and let me know how it goes.

**Comment by dyu** [comment]
Thanks @Bogdan-Vasile 
I'll take a look and see where the fix can be applied.

**Comment by dyu** [comment]
The only fix needed is the one you initially submitted for IdStrategy.

For circular data, the solution is to basically tag your pojos and reserve the single-digit numbers (1-9) as the tag for the fields that are part of the hashCode computation.

This needs to be documented for data that contains hashset and has cyclic refs.

**Comment by dyu** [comment]
Fixed since 1.5.9 and 1.6.1
